Add cabinet API

qmfrederik opened the pull request:
- FDICreate
- FDICopy
- FDIDestroy

qmfrederik committed:
Add Cabinet API
- FDICreate
- FDICopy
- FDIDestroy

qmfrederik committed:
Initialize erf

AArnott commented on src/Cabinet/Cabinet+CpuType.cs:
```diff
@@ -0,0 +1,36 @@
+        /// In the 16-bit version of FDI, specifies the CPU type and can be any of the following values.
+        /// </summary>
+        [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1300:Element must begin with upper-case letter", Justification = "Native naming convention")]
+
```
why the blank line?

AArnott commented on src/Cabinet/Cabinet.cs:
```diff
@@ -0,0 +1,207 @@
+        /// If the function succeeds, it returns a non-<see cref="IntPtr.Zero"/> HFDI context pointer; otherwise, it returns <see cref="IntPtr.Zero"/>.
+        /// </returns>
+        [DllImport(nameof(Cabinet), CallingConvention = CallingConvention.Cdecl)]
+        public static unsafe extern FdiHandle FDICreate(FNALLOC pfnalloc, FNFREE pfnfree, FNOPEN pfnopen, FNREAD pfnread, FNWRITE pfnwrite, FNCLOSE pfnclose, FNSEEK pfnseek, CpuType cpuType, ERF* perf);
```
Please add `[Friendly(FriendlyFlags.Out)]` to the `perf` parameter so that an overload with `out ERF` will be generated.
Also: Can you put one parameter per line?

AArnott commented on src/Cabinet/Cabinet.cs:
```diff
@@ -0,0 +1,207 @@
+        /// </returns>
+        [DllImport(nameof(Cabinet), CallingConvention = CallingConvention.Cdecl)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool FDIDestroy(IntPtr hfdi);
```
We tend to make the handle-releasing p/invoke private since we hand out `SafeHandle` that should be used to close the handle in the end.

AArnott commented on src/Cabinet/Cabinet.cs:
```diff
@@ -0,0 +1,207 @@
+        /// </returns>
+        [DllImport(nameof(Cabinet), CallingConvention = CallingConvention.Cdecl)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool FDICopy(FdiHandle hfdi, string pszCabinet, string pszCabPath, int flags, PFNNOTIFY pfnfdin, IntPtr pfnfdid, IntPtr pvUser);
```
Please use pointers rather than `IntPtr`, as the `IntPtr` versions will be auto-generated.

AArnott commented on src/Kernel32/Kernel32.cs:
```diff
@@ -375,6 +376,35 @@ public static partial class Kernel32
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool SetFilePointer(
+            SafeObjectHandle hFile,
+            long liDistanceToMove,
+            out long lpNewFilePointer,
```
`LONG` in native code is still just a 32-bit integer. When native code _really_ means 64-bit they use `LONGLONG`.  So I suspect this fails until you change it to `int`.

qmfrederik replied:
Thanks for catching this, the P/Invoke declaration was for `SetFilePointerEx` instead of `SetFilePointer`

AArnott commented on src/PInvoke.sln:
```diff
@@ -142,6 +142,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Ole32", "Ole32\Ole32.csproj
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Ole32.Tests", "Ole32.Tests\Ole32.Tests.csproj", "{8B01C254-2EF7-4724-AFD7-FAD39BE69618}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Cabinet", "Cabinet\Cabinet.csproj", "{966BE4C3-657F-4D2C-8717-C04804064186}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Cabinet.Tests", "Cabinet.Tests\Cabinet.Tests.csproj", "{12C7FC54-61AA-413E-A53F-B7AD5D738C70}"
+EndProject
```
Good. But don't forget the [rest of these steps]

https://github.com/dotnet/pinvoke/blob/b8977c166596fa8e27df510a77927324a48d2907/templates/AddNewLibrary.ps1#L53-L55

AArnott reviewed (commented):
I see your PR is a draft so perhaps I'm jumping the gun, but here is some feedback.
Thanks so much for contributing. This is a lot of good stuff!

qmfrederik committed:
PR feedback

qmfrederik commented:
@AArnott Thanks for the review! I addressed your comments. I mainly marked the PR as draft because there are a bunch of other PRs open and I didn't want to flood you with PRs 😄 .

AArnott commented on README.md:
```diff
@@ -95,6 +95,7 @@ user32.dll   |`PInvoke.User32`  | [![NuGet](https://buildstats.info/nuget/PInvok
 userenv.dll  |`PInvoke.Userenv` | [![NuGet](https://buildstats.info/nuget/PInvoke.Userenv)](https://www.nuget.org/packages/PInvoke.Userenv)|Windows User Environment
 uxtheme.dll  |`PInvoke.UxTheme` | [![NuGet](https://buildstats.info/nuget/PInvoke.UxTheme)](https://www.nuget.org/packages/PInvoke.UxTheme)|[Windows Visual Styles][UxTheme]
 WtsApi32.dll |`PInvoke.WtsApi32`| [![NuGet](https://buildstats.info/nuget/PInvoke.WtsApi32)](https://www.nuget.org/packages/PInvoke.WtsApi32)|[Windows Remote Desktop Services][WtsApi32]
+cabinet.dll  |`PInvoke.Cabinet` | [![NuGet](https://buildstats.info/nuget/PInvoke.Cabinet)](https://www.nuget.org/packages/PInvoke.Cabinet)|[Cabinet API Functions][Cabinet]
```
Please keep the table sorted. (if there's a pre-existing sorting error, feel free to fix it)

AArnott commented on src/Cabinet/Cabinet.cs:
```diff
@@ -0,0 +1,224 @@
+        public delegate uint FNSEEK(int hf, int dist, SeekOrigin seektype);
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        public delegate int PFNNOTIFY(NOTIFICATIONTYPE fdint, NOTIFICATION fdin);
```
You've done a great job with docs everywhere else. Is this one missing docs because no docs exist, or was it an oversight? This wouldn't be a reason to deny merging a PR, but I wanted to give you a chance to fill it in.

AArnott commented on src/Cabinet/Cabinet.cs:
```diff
@@ -0,0 +1,224 @@
+            FNCLOSE pfnclose,
+            FNSEEK pfnseek,
+            CpuType cpuType,
+            [Friendly(FriendlyFlags.Out)]
+            ERF* perf);
```
One line per parameter, please.

AArnott reviewed (changes requested):
Very close

qmfrederik committed:
API feedback

AArnott reviewed (approved):
Thanks

qmfrederik committed:
Build fixes

qmfrederik committed:
PR feedback
